### PR TITLE
Add Live Reload

### LINKED ISSUE #33: Demonstrate live reload without watchexec
Does this implement paketo-buildpacks/php#542 ?

### PR DESCRIPTION
**Pull request by joshuatcasey**
Add live reload

Resolves #33 

## Checklist
* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by joshuatcasey** [comment]
Blocked on a release of https://github.com/paketo-buildpacks/libreload-packit

File: README.md
Comment (sophiewigmore) [inline README.md:45]: Are you planning to add anything here about the live reload?

File: integration/httpd_reload_test.go
Comment (sophiewigmore) [inline integration/httpd_reload_test.go:35]: and `BP_LIVE_RELOAD_ENABLED` set to true

**Comment by sophiewigmore** [comment]
The changes here look reasonable to me, but I'd like to be able to test it out locally when a release is available of libreload-packit

File: README.md
Comment (joshuatcasey) [reply README.md:45]: Good call, pushed a change to README

File: integration/httpd_reload_test.go
Comment (joshuatcasey) [reply integration/httpd_reload_test.go]: done

**Comment by joshuatcasey** [comment]
> The changes here look reasonable to me, but I'd like to be able to test it out locally when a release is available of libreload-packit

Sounds good. It can be tested locally with the prerelease version of `libreload-packit` which is why this PR isn't in `Draft` but of course we need to wait.

Another thought I had - do we have confidence that we're not leaking processes by reloading `procmgr-binary`? I'll set some time aside (probably tomorrow) to manually inspect the process list.

**Comment by sophiewigmore** [comment]
@joshuatcasey what do you mean by leaking processes?

**Comment by joshuatcasey** [comment]
> @joshuatcasey what do you mean by leaking processes?

`procmgr-binary` kicks off `fpm` and `nginx|httpd`. If `watchexec` restarts `procmgr-binary`, I'm curious if those `fpm` and `nginx|httpd` processes are cleaned up correctly and new processes created.

**Comment by joshuatcasey** [comment]
@sophiewigmore while investigating the "leaking processes" above, I had a brainwave: do we really even need `watchexec` here? Won't `httpd` and `nginx` just pick up file changes?

And it looks like that answer is.... Yes. I'll remove `watchexec` and `libreload-packit` from this PR shortly.

**Comment by joshuatcasey** [comment]
I've force-pushed this branch so that it only contains `README` and integration test changes. 

The original "working" PR is over on branch `jtc/add-watchexec` but we don't need to save that once this PR is merged.

**Review by TisVictress — APPROVED** [review]

File: integration/nginx_reload_test.go
Comment (sophiewigmore) [inline integration/nginx_reload_test.go:35]: @joshuatcasey I'm confused about this. `BP_LIVE_RELOAD_ENABLED` is not set to true - does this line need to be modified? Or should `BP_LIVE_RELOAD_ENABLED` be set?

I am assuming just this line needs to change?

**Comment by sophiewigmore** [comment]
@TisVictress why was the `BP_LIVE_RELOAD_ENABLED` added to the HTTPD/Nginx buildpacks if that functionality works without the watchexec buildpack?

**Comment by TisVictress** [comment]
> @TisVictress why was the `BP_LIVE_RELOAD_ENABLED` added to the HTTPD/Nginx buildpacks if that functionality works without the watchexec buildpack?

I needed to make this change in the nginx and httpd buildpacks so that I could add support for watchexec to the [web-servers] buildpack(https://github.com/paketo-buildpacks/web-servers/commit/38c3d1b43ac1a4fa7850542c531fe078972f9365)

**Comment by sophiewigmore** [comment]
@TisVictress but my question is why is that buildpack needed if this functionality works without watchexec?

**Comment by sophiewigmore** [comment]
Current state of this PR:
- I'd like to understand if the `php-fpm` start command part of this buildpack need to be reloaded. In the current implementation, this is not a reloadable process type as far as I know. 

- Even though as seen in this PR, watchexec isn't needed, should we include it for parity with what the Nginx and HTTPD buildpacks do? It seems silly to pull in the watchexec buildpack for no reason.

-

File: integration/nginx_reload_test.go
Comment (sophiewigmore) [reply integration/nginx_reload_test.go:35]: @joshuatcasey this still needs to change since the env var isn't set.

**Comment by joshuatcasey** [comment]
@sophiewigmore @TisVictress I've updated this PR to include web server config live reload.

Note - the served files are automatically reloaded by `httpd` or `nginx`, no additional configuration is needed for this.

The conf files are live reloaded by watching the appropriate directory.

I felt this was a nice division of responsibility between the buildpack and the web server. Let the web server do its normal thing, which is likely faster and less intrusive than force-restarting the whole process.

File: build.go
Comment (joshuatcasey) [reply build.go:59]: Yeah. I've gotten into the habit of using `make` because it "just works" in any situation and don't always know if/when it will be used as per later test cases. If there's a preference the other way I'll use that convention.

**Comment by sophiewigmore** [comment]
@joshuatcasey love to see libreload-packit in action! 👏 

I'm wondering if the FPM configuration should be added to the watch path as well? Did you look into this?

**Comment by joshuatcasey** [comment]
> I'm wondering if the FPM configuration should be added to the watch path as well? Did you look into this? https://github.com/paketo-buildpacks/php-fpm#fpm-configuration-sources has the configuration sources, although I think we could get away with just watching the user-provided configuration under /workspace/.php.fpm.d/*.conf

Short answer: No

Long answer: Looks like we could just watch that one dir and it would "just work" but I don't really know much about how FPM works. Can a `Process Manager` be used to do the same thing as watchexec? Which processes is it managing? I just don't have much knowledge about it.

**Comment by sophiewigmore** [comment]
@joshuatcasey I'm not comfortable with calling this issue done until that has been investigated, I don't know off of the top of my head. It seems entirely reasonable to me that we would include the user-provided configuration path as a watch path, and the process restarts on changes. From a preliminary search:
-  I see that there is a `php-fpm restart` command that will stop/start service with configuration changes, which affects the master FPM process.
- There is a `php-fpm reload` command doesn't stop/start service, but re-reads configuration
I'm wondering what the behaviour looks like if watchexec restarts the process. Regardless, it seems like a natural / expected type of behaviour for PHP FPM given that `reload` exists.

**Comment by joshuatcasey** [comment]
@sophiewigmore Thanks for the context - I'll spend some time this afternoon looking into FPM, especially if there's any way we could integration test FPM conf changes.

**Comment by sophiewigmore** [comment]
The default app we use in the PHP FPM buildpack integration test has a empty `.php.fpm.d/user.conf`. When FPM is configured, the ["base" configuration ](https://github.com/paketo-buildpacks/php-fpm/blob/main/config/php-fpm-base.conf) provided by the buildpack gets overridden by any user-provided configuration. You could probably change one of the settings from the base conf file in the user conf. file. 

Maybe after an initial build with a blank user-conf file, you could add something like:
```
[www]
; The number of child processes to be created when pm is set to 'static' and the
; maximum number of child processes when pm is set to 'dynamic' or 'ondemand'.

pm.max_children = 1
```
This would change one of the settings in the FPM configuration (this specific field is set to 5 in the "base" config I linked), and hopefully the change would be picked up? You can also see if FPM logs anything about changes. Not sure exactly what would be the best configuration to mess with, but something along those lines perhaps?

**Comment by joshuatcasey** [comment]
@sophiewigmore those look like good ideas for integration testing.

It looks like all of `fpm`, `nginx`, and `httpd` will respond to a `HUP` signal to reload their configuration without a total restart. Also, `watchexec` can issue a signal to a process without a hard restart. I like that this seems to be the most graceful way to prompt a config reload, and seems much more natural than issuing a hard-restart. 

Doing this may require a change to `libreload-packit` so watch out for that if it comes.

**Comment by sophiewigmore** [comment]
Is it worth adding to libreload-packit if the use case is a one-off? The PHP Start buildpack is a pretty unique case

**Comment by joshuatcasey** [comment]
@paketo-buildpacks/php-maintainers Updated. Note the checks for directory existence before watching them - `watchexec` will hard fail if the directory does not exist.

File: build.go
Comment (sophiewigmore) [inline build.go:80]: I feel like this is misleading since the server itself is reloadable

File: build.go
Comment (sophiewigmore) [inline build.go:77]: What do these flags mean? Some of them I realize are typical arguments, but I don't think I've seen `--on-busy-update` or `--signal` before

File: build.go
Comment (sophiewigmore) [reply build.go:80]: I don't really think this is a necessary thing to log. Maybe as a debug log if you really wanted to

File: build.go
Comment (sophiewigmore) [inline build.go]: Why add this path to the watch path in the case where this directory doesn't exist? Shouldn't this only happen in the `shouldEnableReload && exists` case?

File: build.go
Comment (sophiewigmore) [inline build.go:136]: Why doesn't the FPM conf path get added to the watch path?

File: build.go
Comment (sophiewigmore) [inline build.go]: Can we delete this?

Comment (sophiewigmore) [reply build.go]: But also, where are the watch paths used in the code? I see where they're getting updated but not used.

Comment (sophiewigmore) [reply build.go]: oops I think this is an out of date comment

File: integration/ServeHeader.go
Comment (sophiewigmore) [inline integration/ServeHeader.go:17]: WHy is this neeed?

Comment (sophiewigmore) [inline integration/ServeHeader.go:17]: Why is this needed over what you added in https://github.com/paketo-buildpacks/occam/pull/169?

File: build_test.go
Comment (sophiewigmore) [inline build_test.go:169]: I think the reloadable process should be the default?

File: build.go
Comment (sophiewigmore) [inline build.go:167]: Let me make sure I understand:
- if live reload is enabled, the `procs.yml` will contain the reloadable processes (instead of the original)
- the default process type is still `web`, but it contains the watchexec processes?

In most cases we have:
- a separate process for reloadable process types (`Type: "reload-web"`), which is the `default` process
- still have the regular `web` process, but it is NOT the `default`.

Do you think it's possible to mimic that behaviour? If you considered that, what is the benefit of doing this this way?

File: build.go
Comment (joshuatcasey) [reply build.go:80]: Fair point. HTTDP will still reload your changed application files.

How about this message at debug? `HTTPD configuration will not be reloadable since .httpd.conf.d folder not found`

File: build.go
Comment (joshuatcasey) [reply build.go:77]: * `--on-busy-update=signal` means issue a signal to the process that watchexec owns when something watched has changed
* `--signal=SIGHUP` means to issue the hangup/HUP specifically

Normally we use `--restart` since we do want the entire process restarted in most cases. The three processes we care about (`nginx`, `httpd`, and `fpm`) all respond to a `HUP` signal by reloading configuration.

* `nginx`: http://nginx.org/en/docs/control.html
* `httpd`: https://httpd.apache.org/docs/2.4/stopping.html
* `fpm`: reference TBD

```shell
% watchexec --help
watchexec 1.20.4
Execute commands when watched files change

USAGE:
    watchexec [OPTIONS] <command>...

...

Behaviour options:
        --on-busy-update <on-busy-update>    Select the behaviour to use when receiving events while the command is running. Current default is queue, will change to do-nothing in 2.0. [possible
                                             values: do-nothing, queue, restart, signal]
    -r, --restart                            Restart the process if it's still running. Shorthand for --on-busy-update=restart
    -s, --signal <signal>                    Specify the signal to send when using --on-busy-update=signal
```

Comment (joshuatcasey) [reply build.go:77]: I have to run an errand. When I get back I will update this PR to fix the integration test and update since `main` was fixed, and respond to the various comments.

File: build.go
Comment (sophiewigmore) [reply build.go:80]: sounds great

File: build.go
Comment (joshuatcasey) [reply build.go]: I chose to leave this process definition the same as before since according to the lifecycle, it hasn't changed.

This buildpack has a pretty unique process structure where the entrypoint is another binary that forks out other processes. The `Type` of that entrypoint shouldn't matter as long as it's marked `Default: true`. 

My understanding from https://github.com/buildpacks/spec/blob/main/platform.md is that `Type: web` is only a magic name when API < 0.6, in which case the process with `Type: web` is chosen as the default entrypoint if no other process is marked as default.

Since all of the process reloading is delegated to `procmgr-binary` in this case it made sense to leave the original process definition alone.

**Comment by joshuatcasey** [comment]
Apparently using `Serve(WithHeader)` from occam just doesn't work. Investigating that now.

**Comment by joshuatcasey** [comment]
Blocked on https://github.com/paketo-buildpacks/occam/pull/178

**Comment by sophiewigmore** [comment]
@joshuatcasey I've merged and release your PR in https://github.com/paketo-buildpacks/occam/releases/tag/v0.13.3

**Comment by joshuatcasey** [comment]
Thanks @sophiewigmore this PR is now all green. I'll respond to any remaining questions.

File: build.go
Comment (joshuatcasey) [reply build.go:77]: Hmmm.... I can't seem to find a reference that FPM will reload configuration when it receives the `HUP` signal. It does seem to react to the `USR2` signal though. https://linux.die.net/man/8/php-fpm

This open issue from 2014 would indicate that FPM does not react to the `HUP` signal. https://bugs.php.net/bug.php?id=67553.

I will see what manual testing I can do.

Comment (sophiewigmore) [reply build.go:77]: Ah great work discovering that FPM catch. I'll take a look as well

File: build.go
Comment (sophiewigmore) [reply build.go]: That seems reasonable to me, thank you for explaining this!

File: build.go
Comment (sophiewigmore) [reply build.go:77]: Alright testing FPM (the web-server reload on config changes work flawlessly :D):
- Built the HTTPD reload test app with the reloadable process env var set.
- Created a `user.conf` file with absurd configuration:
```
; User-set FPM configuration
[www]
; The number of child processes to be created when pm is set to 'static' and the
; maximum number of child processes when pm is set to 'dynamic' or 'ondemand'.

pm.max_children = 2000
```
- copied it into the running container under `/workspace/.php.fpm.d/user.conf`
- I saw no logs from the container about changes to FPM configuration
- I was still able to curl the container as normal
- I deleted the `user.conf` file from `.php.fpm.d/`
- I saw logs from the container about FPM failing to initialize:
```
[07-Oct-2022 19:21:07] WARNING: Nothing matches the include pattern '/workspace/.php.fpm.bp/*.conf' from /layers/paketo-buildpacks_php-fpm/php-fpm-config/base.conf at line 30.
[07-Oct-2022 19:21:07] WARNING: Nothing matches the include pattern '/workspace/.php.fpm.d/*.conf' from /layers/paketo-buildpacks_php-fpm/php-fpm-config/base.conf at line 32.
[07-Oct-2022 19:21:07] NOTICE: [pool www] 'user' directive is ignored when FPM is not running as root
[07-Oct-2022 19:21:07] NOTICE: [pool www] 'group' directive is ignored when FPM is not running as root
[07-Oct-2022 19:21:07] ERROR: unable to bind listening socket for address '127.0.0.1:9000': Address already in use (98)
[07-Oct-2022 19:21:07] ERROR: FPM initialization failed
[[Command exited with 78]]
```
- I still see an FPM process running on the container (the original one i think)
- I can still curl the container endpoint no problem

Comment (sophiewigmore) [reply build.go:77]: For test 2:
- I modified `  "--signal", "SIGHUP",` in the FPM block to be ` "--signal", "SIGUSR2",` per your recommendation and rebuilt the buildpack
- Did the same steps as before
- Copying the `user.conf` file in caused the container to only spit out:
```
[[Command killed by User2]]
```
- It appears that FPM is still running on the container and I can access the container endpoint
- VERY interestingly, if I copy in the file again (that is already on the container file system) I get:
```
[07-Oct-2022 19:43:25] NOTICE: [pool www] 'user' directive is ignored when FPM is not running as root
[07-Oct-2022 19:43:25] NOTICE: [pool www] 'group' directive is ignored when FPM is not running as root
[07-Oct-2022 19:43:25] ERROR: unable to bind listening socket for address '127.0.0.1:9000': Address already in use (98)
[07-Oct-2022 19:43:25] ERROR: FPM initialization failed
[[Command exited with 78]]
```
- The original warning that there's nothing inside of `.php.fpm.d` is gone, so FPM has realized we've added something
- It can't restart FPM it seems because the binding socket is still intact

Comment (sophiewigmore) [reply build.go:77]: So, essentially with both signal types issued in the watchexec process, neither one is able to kill the connection fully and reload

Comment (sophiewigmore) [reply build.go:77]: Need to do some more investigation

**Review by TisVictress — DISMISSED** [review]

**Comment by sophiewigmore** [comment]
@TisVictress this PR is still in progress!

**Comment by joshuatcasey** [comment]
I think it looks good as of `8d026a0c74b375d4dfed5331cf0329980eac152d`!

**Comment by joshuatcasey** [comment]
@paketo-buildpacks/php-maintainers Are we good to merge this?

**Review by sophiewigmore — DISMISSED** [review]

**Review by thitch97 — APPROVED** [review]

File: build.go
Comment (thitch97) [reply build.go]: > So, essentially with both signal types issued in the watchexec process, neither one is able to kill the connection fully and reload

What was the issue here? Not going to block on this since the tests pass & I see that they check for successful reload of the procs but just wondering what the resolution was, if any? Did you observe this as well, @joshuatcasey?

Comment (joshuatcasey) [reply build.go]: @sophiewigmore and I paired to check this, we were able to get the expected behavior in that FPM reloads configuration when it receives the `SIGUSR2` signal.

**Review by thitch97 — APPROVED** [review]

**Pull request merged**
